spec/requests/end_to_end_spec.rb

```diff
@@ -1,215 +0,0 @@
-        click_on "Sign in"
-
-        expect(page).to have_content("restricted kablooie")
-
-        visit "http://localhost:4567/users/sign_out"
-
-        Timecop.travel(Time.now.utc + GDS::SSO::Config.auth_valid_for + 5.minutes) do
-          visit "http://#{@client_host}/restricted"
-        end
-
-        expect(page).to have_content("Sign in")
-      end
-
-      it "should accept signon's remembered authentication N hours after login" do
-        visit "http://#{@client_host}/restricted"
-        expect(page).to have_content("Sign in")
-        fill_in "Email", with: "test@example-client.com"
-        fill_in "Password", with: "q1w2e3r4t5y6u7i8o9p0"
-        click_on "Sign in"
-
-        expect(page).to have_content("restricted kablooie")
-
-        Timecop.travel(Time.now.utc + GDS::SSO::Config.auth_valid_for + 5.minutes) do
-          visit "http://#{@client_host}/restricted"
-        end
-
-        expect(page).to have_content("restricted kablooie")
-      end
-
-      it "should not require re-authentication with signon fewer than N hours after login" do
-        visit "http://#{@client_host}/restricted"
-        expect(page).to have_content("Sign in")
-        fill_in "Email", with: "test@example-client.com"
-        fill_in "Password", with: "q1w2e3r4t5y6u7i8o9p0"
-        click_on "Sign in"
-
-        expect(page).to have_content("restricted kablooie")
-
-        Timecop.travel(Time.now.utc + GDS::SSO::Config.auth_valid_for - 5.minutes) do
-          visit "http://#{@client_host}/restricted"
-        end
-
-        expect(page).to have_content("restricted kablooie")
-      end
-    end
-  end
-
-  describe "OAuth based API client accesses" do
-    before :each do
-      page.driver.header "accept", "application/json"
-      authorize_signon_api_user
-
-      token = "caaeb53be5c7277fb0ef158181bfd1537b57f9e3b83eb795be3cd0af6e118b28"
-      page.driver.header "authorization", "Bearer #{token}"
-    end
-
-    specify "access to a restricted page for an api client requires auth" do
-      page.driver.header "authorization", "Bearer Bad Token"
-      visit "http://#{@client_host}/restricted"
-      expect(page.driver.response.status).to eq(401)
-    end
-
-    specify "setting a correct bearer token allows sign in" do
-      visit "http://#{@client_host}/restricted"
-      expect(page).to have_content("restricted kablooie")
-    end
-
-    specify "setting a correct bearer token picks up permissions" do
-      visit "http://#{@client_host}/this_requires_signin_permission"
-      expect(page).to have_content("you have signin permission")
-    end
-
-    specify "a token for one app cannot be used to access a different app" do
-      page.driver.header "authorization", "Bearer 98c72f4da02fdc43398e029d05567542944d2a9b0df3c20b0accd8bd6c5dc728"
-      visit "http://#{@client_host}/restricted"
-      expect(page.driver.response.status).to eq(401)
-    end
-  end
-
-  context "when in api_only mode" do
```

Comment:
Just comparing the new tests we have versus the ones we're getting rid of - is it worth translating the api_only ones or is there a reason they're unecessary now?

Reply:
Cool, I think in general this PR looks good to me (and what a great bit of tech debt you're getting rid of!!), I think my only questions were about whether we have all the tests still, so maybe when you're done you could talk me through the two files and reasoning for any tests that we've dropped, then I'd be happy to ✅

Reply:
Thanks @BeckaL, I've added the missing test in [ce63dc2](https://github.com/alphagov/gds-sso/pull/274/commits/ce63dc2d323265ed3408f0e7c83f013d3bcafaf2)

Here's a breakdown of the test changes:

["Web client accesses, a non-restricted page can be accessed without authentication"](https://github.com/alphagov/gds-sso/blob/1eb1feaf97dd3318d57a0a465e3797521b25c0f9/spec/requests/end_to_end_spec.rb#L26) is covered by: ["when accessing a route that doesn't require permissions or authentication, it allows access"](https://github.com/alphagov/gds-sso/pull/274/files#diff-50bfe25c158c07cc493853c799f98e97e60b61046850d3eac755e6dbae8322abR5)

["Web client accesses, access to a restricted page for an approved application requires only authentication"](https://github.com/alphagov/gds-sso/blob/1eb1feaf97dd3318d57a0a465e3797521b25c0f9/spec/requests/end_to_end_spec.rb#L31) is covered by: ["when accessing a route that requires authentication, it allows access for an authenticated user"](https://github.com/alphagov/gds-sso/pull/274/files#diff-50bfe25c158c07cc493853c799f98e97e60b61046850d3eac755e6dbae8322abR21)

["Web client accesses, access to a restricted page for an approved application requires only authentication"](https://github.com/alphagov/gds-sso/blob/1eb1feaf97dd3318d57a0a465e3797521b25c0f9/spec/requests/end_to_end_spec.rb#L41) does not have an analogue test, it did not test any distinct change in functionality of this gem nor is it clear that the scenario explained is actually tested.

["Web client accesses, access to a page that requires signin permission granted"](https://github.com/alphagov/gds-sso/blob/1eb1feaf97dd3318d57a0a465e3797521b25c0f9/spec/requests/end_to_end_spec.rb#L62) has a replacement test in ["when accessing a route that requires a permission, it allows access when an authenticated user has the permission"](https://github.com/alphagov/gds-sso/pull/274/files#diff-50bfe25c158c07cc493853c799f98e97e60b61046850d3eac755e6dbae8322abR73) - the previous test was broken as the signin permission is implicit as I recall, if that happens again it should be exposed by: ["when accessing a route that requires a permission, it restricts access when an authenticated user lacks the permission"](https://github.com/alphagov/gds-sso/pull/274/files#diff-50bfe25c158c07cc493853c799f98e97e60b61046850d3eac755e6dbae8322abR79)

["Web client accesses, remotely signed out, should prevent all access to the application until successful signin"](https://github.com/alphagov/gds-sso/blob/1eb1feaf97dd3318d57a0a465e3797521b25c0f9/spec/requests/end_to_end_spec.rb#L84) is covered by ["when accessing a route that requires authentication, it restricts access if a user is authenticated but remotely signed out"](https://github.com/alphagov/gds-sso/pull/274/files#diff-50bfe25c158c07cc493853c799f98e97e60b61046850d3eac755e6dbae8322abR28)

["Web client accesses, session expiry, it should force you to re-authenticate with signon N hours after login"](https://github.com/alphagov/gds-sso/blob/1eb1feaf97dd3318d57a0a465e3797521b25c0f9/spec/requests/end_to_end_spec.rb#L117) does not have a direct replacement because it seems a bit dodgy, the naming of it implies it's time related but then there is a visit to sign out link used so I don't believe it's actually testing the time.

["Web client accesses, it should accept signon's remembered authentication N hours after login"](https://github.com/alphagov/gds-sso/blob/1eb1feaf97dd3318d57a0a465e3797521b25c0f9/spec/requests/end_to_end_spec.rb#L135) seems a bit dodgy too as it seems like it should be actually testing that the session expires and you need to be reauthenticated with signon. However this test doesn't assert that and would seemingly pass if the time expiring had no effect. A replacement test is added in ["when accessing a route that requires authentication, it restricts access if a user is authenticated but session has expired"](https://github.com/alphagov/gds-sso/pull/274/files#diff-50bfe25c158c07cc493853c799f98e97e60b61046850d3eac755e6dbae8322abR37)

["Web client accesses, it should not require re-authentication with signon fewer than N hours after login"](https://github.com/alphagov/gds-sso/blob/1eb1feaf97dd3318d57a0a465e3797521b25c0f9/spec/requests/end_to_end_spec.rb#L151) doesn't have a dedicated companion test as it's exercising the same logic as ["when accessing a route that requires authentication, it allows access for an authenticated user"](https://github.com/alphagov/gds-sso/pull/274/files#diff-50bfe25c158c07cc493853c799f98e97e60b61046850d3eac755e6dbae8322abR21).

["OAuth based API client accesses, access to a restricted page for an api client requires auth"](https://github.com/alphagov/gds-sso/blob/1eb1feaf97dd3318d57a0a465e3797521b25c0f9/spec/requests/end_to_end_spec.rb#L178) is covered by ["when accessing a route that requires authentication, it restricts access when given an invalid bearer token"](https://github.com/alphagov/gds-sso/pull/274/files#diff-50bfe25c158c07cc493853c799f98e97e60b61046850d3eac755e6dbae8322abR55)

["OAuth based API client accesses, setting a correct bearer token allows sign in"](https://github.com/alphagov/gds-sso/blob/1eb1feaf97dd3318d57a0a465e3797521b25c0f9/spec/requests/end_to_end_spec.rb#L184) is covered by ["when accessing a route that requires authentication, it allows access when given a valid bearer token"](https://github.com/alphagov/gds-sso/pull/274/files#diff-50bfe25c158c07cc493853c799f98e97e60b61046850d3eac755e6dbae8322abR47)

["OAuth based API client accesses, setting a correct bearer token picks up permissions"](https://github.com/alphagov/gds-sso/blob/1eb1feaf97dd3318d57a0a465e3797521b25c0f9/spec/requests/end_to_end_spec.rb#L189) does not have a new test. This doesn't seem to a distinct enough scenario from the existing permission test to warrant another test.

["OAuth based API client accesses, a token for one app cannot be used to access a different app"](https://github.com/alphagov/gds-sso/blob/1eb1feaf97dd3318d57a0a465e3797521b25c0f9/spec/requests/end_to_end_spec.rb#L194) does not have a new test. This is a signon concern.

["when in api_only mode, accessing without a bearer token is not authorized"](https://github.com/alphagov/gds-sso/blob/1eb1feaf97dd3318d57a0a465e3797521b25c0f9/spec/requests/end_to_end_spec.rb#L210) is covered by ["when accessing a route that requires authentication, it returns a 401 when a bearer token is missing and the app is api_only"](https://github.com/alphagov/gds-sso/pull/274/files#diff-50bfe25c158c07cc493853c799f98e97e60b61046850d3eac755e6dbae8322abR64)

Reply:
Thank you so much for this, super helpful!